### DISCUSSION TITLE
Pre-rendering and Data Fetching question?

### DISCUSSION THREAD

**Discussion by yangyangnote** [vercel/next.js#14578] [Help]
https://nextjs.org/learn/basics/data-fetching/implement-getstaticprops
When I follow this tutorial to this point, I always get an error
![截屏2020-06-26 下午2 12 28](https://user-images.githubusercontent.com/19338570/85826383-15270b80-b7b7-11ea-91f9-ca7b451dfbde.png)
TypeError: Cannot read property 'map' of undefined
Here are the code involved

lib/posts.js
```
import fs from 'fs'
import path from 'path'
import matter from 'gray-matter'

const postsDirectory = path.join(process.cwd(), 'posts')

export function getSortedPostsData() {
  // Get file names under /posts
  const fileNames = fs.readdirSync(postsDirectory)
  const allPostsData = fileNames.map(fileName => {
    // Remove ".md" from file name to get id
    const id = fileName.replace(/\.md$/, '')

    // Read markdown file as string
    const fullPath = path.join(postsDirectory, fileName)
    const fileContents = fs.readFileSync(fullPath, 'utf8')

    // Use gray-matter to parse the post metadata section
    const matterResult = matter(fileContents)

    // Combine the data with the id
    return {
      id,
      ...matterResult.data
    }
  })
  // Sort posts by date
  return allPostsData.sort((a, b) => {
    if (a.date < b.date) {
      return 1
    } else {
      return -1
    }
  })
}
```

pages/index.js
```
import Head from 'next/head'
import Layout, { siteTitle } from '../components/layout'
import utilStyles from '../styles/utils.module.css'
import { getSortedPostsData } from '../lib/posts'


export async function getStaticProps() {
  const allPostsData = getSortedPostsData()
  return {
    props: {
      allPostsData
    }
  }
}


export default function Home ({ allPostsData }) {
  return (
    <Layout home>
      <Head>
        <title>{siteTitle}</title>
      </Head>
      <section className={utilStyles.headingMd}>
        <p>[Hello, I'm Yang. I'm a iOS enginer.]</p>
        <p>
          (This is a sample website - you’ll be building a site like this on{' '}
          <a href="https://nextjs.org/learn">our Next.js tutorial</a>.)
        </p>
      </section>
      <section className={`${utilStyles.headingMd} ${utilStyles.padding1px}`}>
        <h2 className={utilStyles.headingLg}>Blog</h2>
        <ul className={utilStyles.list}>
          {allPostsData.map(({ id, date, title }) => (
            <li className={utilStyles.listItem} key={id}>
              {title}
              <br />
              {id}
              <br />
              {date}
            </li>
          ))}
        </ul>
      </section>
    </Layout>
  )
  
}

```

**Comment by yangyangnote** [comment]
May be a problem at the file level

**Comment by zeekrey** [comment]
I checked the tutorial and it works for me. You could insert a simple `console.log(JSON.stringify(allPostsData))` in _lib/posts.js_ to see if the markdown files are fetched.

**Reply by kmgdevelopment** [reply]
Did this and got `ReferenceError: allPostsData is not defined`, so what's the issue exactly? My posts are in a top-level `posts` folder, markdown content copy/pasted directly from the tutorial. As far as I know I've followed the tutorial exactly.

**Comment by iborba** [comment]
The same to me

**Comment by johnend** [comment]
In order for this to work for me I needed to restart the development server using `ctrl` + `c` and then re-running `npm run dev`

**Reply by sameev** [reply]
After spending way too long trying to figure out where my code was off, this was the answer that got it working for me. Thanks!

**Reply by emrecoban** [reply]
This is the solution. Thanks! They should fix the doc page.

**Reply by shpoopdy** [reply]
This. Now all I can think of is my parent's saying "Well did you try turning it off and on?"

**Reply by kmgdevelopment** [reply]
This solution did NOT work for me. Copy/pasted the code straight from the tutorial, restarted the server, still getting the `.map` error

**Comment by li45** [comment]
This line :  ``{allPostsData.map(({` id, date, title }) => (``

should be replaced with:  ``{allPostsData.allPostsData.map(({ id, date, title }) => (``

The first "allPostsData" you returned is a Object, witch have "props"  that contains a prop "allPostsData".

see [https://stackoverflow.com/questions/30803168/data-map-is-not-a-function](https://stackoverflow.com/questions/30803168/data-map-is-not-a-function)
